libs/custom-paper-handler/src/driver/driver.ts

```diff
@@ -140,83 +140,110 @@ export async function getPaperHandlerWebDevice(): Promise<
 }
 
 export class PaperHandlerDriver implements PaperHandlerDriverInterface {
-  readonly genericLock = new Lock();
-  readonly realTimeLock = new Lock();
+  // By convention, each public API method must hold this lock before starting
+  readonly publicApiMutex = new Mutex(undefined);
+  readonly webDeviceMutex: Mutex<MinimalWebUsbDevice>;
+
   readonly scannerConfig: ScannerConfig = getDefaultConfig();
-  webDevice: MinimalWebUsbDevice;
 
-  constructor(_webDevice: MinimalWebUsbDevice) {
-    this.webDevice = _webDevice;
+  constructor(webDevice: MinimalWebUsbDevice) {
+    this.webDeviceMutex = new Mutex(webDevice);
   }
 
   async connect(): Promise<void> {
-    await this.webDevice.open();
-    debug('opened web device');
-    await this.webDevice.selectConfiguration(CONFIGURATION_NUMBER);
-    debug(`selected configuration ${CONFIGURATION_NUMBER}`);
-    await this.webDevice.claimInterface(INTERFACE_NUMBER);
-    debug(`claimed usb interface ${INTERFACE_NUMBER}`);
+    return await this.publicApiMutex.withLock(() =>
+      this.webDeviceMutex.withLock(async (webDevice) => {
+        await webDevice.open();
+        debug('opened web device');
+        await webDevice.selectConfiguration(CONFIGURATION_NUMBER);
+        debug(`selected configuration ${CONFIGURATION_NUMBER}`);
+        await webDevice.claimInterface(INTERFACE_NUMBER);
+        debug(`claimed usb interface ${INTERFACE_NUMBER}`);
+      })
+    );
   }
 
   async disconnect(): Promise<void> {
     // closing the web device will fail if we have pending requests, so wait for them
-    await this.genericLock.acquire();
-    this.genericLock.release();
-    await this.realTimeLock.acquire();
-    this.realTimeLock.release();
-
-    // await this.webDevice.releaseInterface(0);
-    // debug('released usb interface');
-    await this.webDevice.close();
-    debug('closed web usb device');
+    await this.publicApiMutex.withLock(() =>
+      this.webDeviceMutex.withLock((webDevice) => webDevice.close())
+    );
   }
 
   /**
-   * Should be private, but exposed for development.
+   * Send commands or data on the generic bulk out endpoint.
+   * Expects the caller to acquire the lock for and pass in the WebUsbDevice.
    */
-  getWebDevice(): MinimalWebUsbDevice {
-    return this.webDevice;
+  transferOutGeneric<T>(
```

Comment:
`transferOutGeneric` and `transferInGeneric` are used in a few different ways by callers, so we leave the locking up to the caller and require the WebUsbDevice to be passed into these functions. The `Mutex` package encourages good locking practices in callers because `.withLock` is called to get a reference to `webDevice`, but that usage is not enforced, so it's possible to regress if a direct reference to `webDevice` is added somewhere.